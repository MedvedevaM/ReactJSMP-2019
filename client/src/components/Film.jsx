import React from 'react';
import { Genre } from './Genre.jsx';

export const Film = (props) => {
    const { poster_path, title, genres, release_date, chooseFilm, id } = props;
    const release_year = release_date ? release_date.slice(0, 4) : null;
    return (
        <div className='film'>
            <img src={poster_path}/>
            <div className='flex'>
                <h3 id={id}>{title}</h3>
                <time className='release-year'>{release_year}</time>
            </div>
            <Genre genres={genres}/>
        </div>
    )
};