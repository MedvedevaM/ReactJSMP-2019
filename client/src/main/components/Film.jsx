import React from "react";
import { Genre } from "./Genre.jsx";

export const Film = props => {
	const { poster_path, title, genres, release_date, id } = props;
	const release_year = release_date ? release_date.slice(0, 4) : null;
	return (
		<div className="film">
			<img data-film-id={id} src={poster_path} alt={title} />
			<div className="flex">
				<h3 data-film-id={id}>{title}</h3>
				<time className="release-year">{release_year}</time>
			</div>
			<Genre genres={genres} />
		</div>
	);
};
