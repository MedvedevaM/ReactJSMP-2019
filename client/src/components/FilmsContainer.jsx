import React from "react";
import { Film } from "./Film.jsx";
import { Genre } from "./Genre.jsx";

export const FilmsContainer = (props) => {
    const { films, quantityOfFilms, chooseFilm, chosenFilm, sortBy, onSortParameterClick } = props;
    if (chosenFilm) {
        const formattedFilms = filterFilmsByGenre(films, chosenFilm.genres).map((film) => <Film key={film.id} chooseFilm={chooseFilm} {...film} />);
        return (
            <>
                <div className="search-results">
                    <div className="container flex">
                        <div className="flex">
                            <p className="genre">Films by</p>
                            <Genre genres={chosenFilm.genres} />
                        </div>
                    </div>
                </div>
                <div onClick={chooseFilm} className="flex films-container">{formattedFilms}</div>
            </>
        )
    }

    const formattedFilms = films.map((film) => <Film key={film.id} chooseFilm={chooseFilm} {...film} />);
    return (
        <>
            <div className="search-results">
                <div className="container flex">
                    <p>{quantityOfFilms} movies found</p>
                    <div onClick={onSortParameterClick} className="flex sorter">
                        <p>Sort by</p>
                        <ul onClick={onSortParameterClick} className="flex sorter">
                            <li id="release date" className={sortBy === "release date" ? "chosen-sort-parameter" : ""}>release date</li>
                            <li id="rating" className={sortBy === "rating" ? "chosen-sort-parameter" : ""}>rating</li>
                        </ul>
                    </div>
                </div>
            </div>
            <div onClick={chooseFilm} className="flex films-container">{formattedFilms}</div>
        </>
    )
};

function filterFilmsByGenre(films, neededGenres) {
    return films.filter(film => {
        return film.genres.some(genre => {
            return neededGenres.some(neededGenre => genre === neededGenre);
        });
    });
}