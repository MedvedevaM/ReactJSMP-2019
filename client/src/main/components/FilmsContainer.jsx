import React from "react";
import { connect } from 'react-redux'
import Film from "./Film.jsx";
import Genre from "./Genre.jsx";
import { getFilteredFilms } from './../store/reducers/reducers';

const FilmsContainer = (props) => {
  const { films, quantityOfFilms, chooseFilm, chosenFilm, sortBy, onSortParameterClick, filteredFilms } = props;
  if (chosenFilm && films) {
    const formattedFilms = filteredFilms.map((film) => <Film key={film.id} chooseFilm={chooseFilm} {...film} />);
    return (
      <>
        <div className="search-results">
          <div className="container flex">
            <div className="flex">
              <p className="genre">Films by</p>
              <Genre genres={chosenFilm.genres} />
            </div>
          </div>
        </div>
        <div onClick={chooseFilm} className="flex films-container">{formattedFilms}</div>
      </>
    )
  }

  const formattedFilms = films ? films.map((film) => <Film key={film.id} chooseFilm={chooseFilm} {...film} />) : null;
  return (
    <>
      <div className="search-results">
        <div className="container flex">
          <p>{quantityOfFilms} movies found</p>
          <div onClick={onSortParameterClick} className="flex sorter">
            <p>Sort by</p>
            <ul onClick={onSortParameterClick} className="flex sorter">
              <li data-sort-parameter="release date" className={sortBy === "release date" ? "chosen-sort-parameter" : ""}>release date</li>
              <li data-sort-parameter="rating" className={sortBy === "rating" ? "chosen-sort-parameter" : ""}>rating</li>
            </ul>
          </div>
        </div>
      </div>
      <div onClick={chooseFilm} className="flex films-container">{formattedFilms}</div>
    </>
  )
};

function mapStateToProps(store) {
  return {
    filteredFilms: getFilteredFilms(store)
  }
}

export default connect(mapStateToProps)(FilmsContainer)