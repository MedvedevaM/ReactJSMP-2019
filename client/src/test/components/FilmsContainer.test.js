import React from "react";
import { shallow, mount } from "enzyme";
import toJson from 'enzyme-to-json';
import { filterFilmsByGenre, FilmsContainer  } from '../../main/components/FilmsContainer.jsx';
import { Film } from '../../main/components/Film.jsx';

describe("rendering of container with films", () => {
	const films = [{
		budget: 16000000,
		genres: ['Science Fiction', 'Action', 'Adventure'],
		id: 185567,
		overview: "As a child, Ali Neuman narrowly escaped being murdered by Inkhata, a militant political party at war with Nelson Mandela's African National Congress. Only he and his mother survived the carnage of those years. But as with many survivors, the psychological scars remain.",
		poster_path: "https://image.tmdb.org/t/p/w500/xg7Dh7mjevDgznqw9JHYccNo9ZQ.jpg",
		release_date: "2013-12-04",
		revenue: 0,
		runtime: 110,
		tagline: "",
		title: "Zulu",
		vote_average: 6.7,
		vote_count: 200
	}, {
		budget: 150000000,
		genres: ['Science Fiction', 'Adventure'],
		id: 269149,
		overview: "Determined to prove herself, Officer Judy Hopps, the first bunny on Zootopia's police force, jumps at the chance to crack her first case - even if it means partnering with scam-artist fox Nick Wilde to solve the mystery.",
		poster_path: "https://image.tmdb.org/t/p/w500/sM33SANp9z6rXW8Itn7NnG1GOEs.jpg",
		release_date: "2016-02-11",
		revenue: 1023784195,
		runtime: 108,
		tagline: "Welcome to the urban jungle.",
		title: "Zootopia",
		vote_average: 7.7,
		vote_count: 6795
	}, {
		budget: 23600000,
		genres: ['Drama', 'Adventure'],
		id: 19908,
		overview: "Columbus has made a habit of running from what scares him. Tallahassee doesn't have fears. If he did, he'd kick their ever-living ass. In a world overrun by zombies, these two are perfectly evolved survivors. But now, they're about to stare down the most terrifying prospect of all: each other.",
		poster_path: "https://image.tmdb.org/t/p/w500/vUzzDpVrab1BOG3ogxhRGfLN94d.jpg",
		release_date: "2009-10-07",
		revenue: 102391382,
		runtime: 88,
		tagline: "This place is so dead",
		title: "Zombieland",
		vote_average: 7.2,
		vote_count: 4509
	}, {
		budget: 65000000,
		genres: ['Science Fiction', 'Action', 'Adventure', 'Drama'],
		id: 1949,
		overview: "The true story of the investigation of the Zodiac Killer, a serial killer who terrified the San Francisco Bay Area, taunting police with his ciphers and letters.  The case becomes an obsession for four men as their lives and careers are built and destroyed by the endless trail of clues.",
		poster_path: "https://image.tmdb.org/t/p/w500/bgLyOROfFQI3FqYL7jQbiaV8lkN.jpg",
		release_date: "2007-03-02",
		revenue: 84785914,
		runtime: 157,
		tagline: "There's more than one way to lose your life to a killer.",
		title: "Zodiac",
		vote_average: 7.3,
		vote_count: 2780
	}, {
		budget: 40000000,
		genres: ['Science Fiction', 'Adventure', 'Fantasy'],
		id: 97630,
		overview: "A chronicle of the decade-long hunt for al-Qaeda terrorist leader Osama bin Laden after the September 2001 attacks, and his death at the hands of the Navy S.E.A.L. Team 6 in May, 2011.",
		poster_path: "https://image.tmdb.org/t/p/w500/yg6IDNucLAEj7E5loTyTnUW2pgb.jpg",
		release_date: "",
		revenue: 132820716,
		runtime: 157,
		tagline: "The Greatest Manhunt in History",
		title: "Zero Dark Thirty",
		vote_average: 6.7,
		vote_count: 2028
	}];
	
	const chosenFilm = {
		budget: 40000000,
		genres: ['Action', 'Drama'],
		id: 97630,
		overview: "A chronicle of the decade-long hunt for al-Qaeda terrorist leader Osama bin Laden after the September 2001 attacks, and his death at the hands of the Navy S.E.A.L. Team 6 in May, 2011.",
		poster_path: "https://image.tmdb.org/t/p/w500/yg6IDNucLAEj7E5loTyTnUW2pgb.jpg",
		release_date: "2012-12-19",
		revenue: 132820716,
		runtime: null,
		tagline: "The Greatest Manhunt in History",
		title: "Zero Dark Thirty",
		vote_average: 6.7,
		vote_count: 2028
	};
	
	it('Check filtering of films by genre', () => {
		expect(filterFilmsByGenre([{
				genres: ['Science Fiction', 'Action', 'Adventure'],
			},
			{
				genres: ['Science Fiction', 'Adventure'],
			},
			{
				genres: ['Drama', 'Adventure'],
			},
			{
				genres: ['Science Fiction', 'Action', 'Adventure', 'Drama'],
			},
			{
				genres: ['Science Fiction', 'Adventure', 'Fantasy'],
			},
		], ['Action', 'Drama'])).toEqual([{
				genres: ['Science Fiction', 'Action', 'Adventure'],
			},
			{
				genres: ['Drama', 'Adventure'],
			},
			{
				genres: ['Science Fiction', 'Action', 'Adventure', 'Drama'],
			},
		]);
	});
	
	it("renders correctly FilmsContainer", () => {
		const filmContainer = shallow( <FilmsContainer films={films} /> );
		expect(toJson(filmContainer)).toMatchSnapshot();
	});
	
	it('Check quantity of rendered films without chosen film', () => {
		const filmContainer = mount( <FilmsContainer films={films} /> );
		expect(filmContainer.find(Film).length).toBe(5);
	});
	
	it('Check quantity of rendered films with chosen film', () => {
		const filmContainer = mount( <FilmsContainer chosenFilm={chosenFilm} films={films} /> );
		expect(filmContainer.find(Film).length).toBe(3);
	});
});


