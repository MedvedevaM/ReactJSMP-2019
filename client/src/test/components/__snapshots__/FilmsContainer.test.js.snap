// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly FilmsContainer 1`] = `
<Fragment>
  <div
    className="search-results"
  >
    <div
      className="container flex"
    >
      <p>
         movies found
      </p>
      <div
        className="flex sorter"
      >
        <p>
          Sort by
        </p>
        <ul
          className="flex sorter"
        >
          <li
            className=""
            data-sort-parameter="release date"
          >
            release date
          </li>
          <li
            className=""
            data-sort-parameter="rating"
          >
            rating
          </li>
        </ul>
      </div>
    </div>
  </div>
  <div
    className="flex films-container"
  >
    <Film
      budget={16000000}
      genres={
        Array [
          "Science Fiction",
          "Action",
          "Adventure",
        ]
      }
      id={185567}
      key="185567"
      overview="As a child, Ali Neuman narrowly escaped being murdered by Inkhata, a militant political party at war with Nelson Mandela's African National Congress. Only he and his mother survived the carnage of those years. But as with many survivors, the psychological scars remain."
      poster_path="https://image.tmdb.org/t/p/w500/xg7Dh7mjevDgznqw9JHYccNo9ZQ.jpg"
      release_date="2013-12-04"
      revenue={0}
      runtime={110}
      tagline=""
      title="Zulu"
      vote_average={6.7}
      vote_count={200}
    />
    <Film
      budget={150000000}
      genres={
        Array [
          "Science Fiction",
          "Adventure",
        ]
      }
      id={269149}
      key="269149"
      overview="Determined to prove herself, Officer Judy Hopps, the first bunny on Zootopia's police force, jumps at the chance to crack her first case - even if it means partnering with scam-artist fox Nick Wilde to solve the mystery."
      poster_path="https://image.tmdb.org/t/p/w500/sM33SANp9z6rXW8Itn7NnG1GOEs.jpg"
      release_date="2016-02-11"
      revenue={1023784195}
      runtime={108}
      tagline="Welcome to the urban jungle."
      title="Zootopia"
      vote_average={7.7}
      vote_count={6795}
    />
    <Film
      budget={23600000}
      genres={
        Array [
          "Drama",
          "Adventure",
        ]
      }
      id={19908}
      key="19908"
      overview="Columbus has made a habit of running from what scares him. Tallahassee doesn't have fears. If he did, he'd kick their ever-living ass. In a world overrun by zombies, these two are perfectly evolved survivors. But now, they're about to stare down the most terrifying prospect of all: each other."
      poster_path="https://image.tmdb.org/t/p/w500/vUzzDpVrab1BOG3ogxhRGfLN94d.jpg"
      release_date="2009-10-07"
      revenue={102391382}
      runtime={88}
      tagline="This place is so dead"
      title="Zombieland"
      vote_average={7.2}
      vote_count={4509}
    />
    <Film
      budget={65000000}
      genres={
        Array [
          "Science Fiction",
          "Action",
          "Adventure",
          "Drama",
        ]
      }
      id={1949}
      key="1949"
      overview="The true story of the investigation of the Zodiac Killer, a serial killer who terrified the San Francisco Bay Area, taunting police with his ciphers and letters.  The case becomes an obsession for four men as their lives and careers are built and destroyed by the endless trail of clues."
      poster_path="https://image.tmdb.org/t/p/w500/bgLyOROfFQI3FqYL7jQbiaV8lkN.jpg"
      release_date="2007-03-02"
      revenue={84785914}
      runtime={157}
      tagline="There's more than one way to lose your life to a killer."
      title="Zodiac"
      vote_average={7.3}
      vote_count={2780}
    />
    <Film
      budget={40000000}
      genres={
        Array [
          "Science Fiction",
          "Adventure",
          "Fantasy",
        ]
      }
      id={97630}
      key="97630"
      overview="A chronicle of the decade-long hunt for al-Qaeda terrorist leader Osama bin Laden after the September 2001 attacks, and his death at the hands of the Navy S.E.A.L. Team 6 in May, 2011."
      poster_path="https://image.tmdb.org/t/p/w500/yg6IDNucLAEj7E5loTyTnUW2pgb.jpg"
      release_date="2012-12-19"
      revenue={132820716}
      runtime={157}
      tagline="The Greatest Manhunt in History"
      title="Zero Dark Thirty"
      vote_average={6.7}
      vote_count={2028}
    />
  </div>
</Fragment>
`;
